README
NAME

    Data::Object::Try

ABSTRACT

    Try Class for Perl 5

SYNOPSIS

      use strict;
      use warnings;
      use routines;
    
      use Data::Object::Try;
    
      my $try = Data::Object::Try->new;
    
      $try->call(fun (@args) {
        # try something
    
        return time;
      });
    
      $try->catch('Example::Exception', fun ($caught) {
        # caught an exception
    
        return;
      });
    
      $try->default(fun ($caught) {
        # catch the uncaught
    
        return;
      });
    
      $try->finally(fun (@args) {
        # always run after try/catch
    
        return;
      });
    
      my @args;
    
      my $result = $try->result(@args);

DESCRIPTION

    This package provides an object-oriented interface for performing
    complex try/catch operations.

ATTRIBUTES

    This package has the following attributes:

 arguments

      arguments(ArrayRef)

    This attribute is read-only, accepts (ArrayRef) values, and is
    optional.

 invocant

      invocant(Object)

    This attribute is read-only, accepts (Object) values, and is optional.

 on_catch

      on_catch(ArrayRef[CodeRef])

    This attribute is read-write, accepts (ArrayRef[CodeRef]) values, and
    is optional.

 on_default

      on_default(CodeRef)

    This attribute is read-write, accepts (CodeRef) values, and is
    optional.

 on_finally

      on_finally(CodeRef)

    This attribute is read-write, accepts (CodeRef) values, and is
    optional.

 on_try

      on_try(CodeRef)

    This attribute is read-write, accepts (CodeRef) values, and is
    optional.

METHODS

    This package implements the following methods:

 call

      call(Str | CodeRef $arg) : Object

    The call method takes a method name or coderef, registers it as the
    tryable routine, and returns the object. When invoked, the callback
    will received an invocant if one was provided to the constructor, the
    default arguments if is $result->[4], 4; is $result->[5], 5; any were
    provided to the constructor, and whatever arguments were provided by
    the invocant. is $result->[4], 4; is $result->[5], 5;

    call example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          return [@args];
        });

 callback

      callback(Str | CodeRef $arg) : CodeRef

    The callback method takes a method name or coderef, and returns a
    coderef for registration. If a coderef is provided this method is
    mostly a passthrough.

    callback example #1

        my $try = Data::Object::Try->new;
      
        $try->callback(fun (@args) {
      
          return [@args];
        });

    callback example #2

        package Example;
      
        use Moo;
      
        fun test(@args) {
      
          return [@args];
        }
      
        package main;
      
        my $try = Data::Object::Try->new(
          invocant => Example->new
        );
      
        $try->callback('test');

 catch

      catch(Str $isa, Str | CodeRef $arg) : Any

    The catch method takes a package or ref name, and when triggered checks
    whether the captured exception is of the type specified and if so
    executes the given callback.

    catch example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          die $try;
        });
      
        $try->catch('Data::Object::Try', fun (@args) {
      
          return [@args];
        });

 default

      default(Str | CodeRef $arg) : Object

    The default method takes a method name or coderef and is triggered if
    no catch conditions match the exception thrown.

    default example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          die $try;
        });
      
        $try->default(fun (@args) {
      
          return [@args];
        });

 execute

      execute(CodeRef $arg, Any @args) : Any

    The execute method takes a coderef and executes it with any given
    arguments. When invoked, the callback will received an invocant if one
    was provided to the constructor, the default arguments if any were
    provided to the constructor, and whatever arguments were passed
    directly to this method.

    execute example #1

        my $try = Data::Object::Try->new(
          invocant => Example->new,
          arguments => [1,2,3]
        );
      
        $try->execute(fun (@args) {
      
          return [@args];
        });

 finally

      finally(Str | CodeRef $arg) : Object

    The finally method takes a package or ref name and always executes the
    callback after a try/catch operation. The return value is ignored. When
    invoked, the callback will received an invocant if one was provided to
    the constructor, the default arguments if any were provided to the
    constructor, and whatever arguments were provided by the invocant.

    finally example #1

        my $try = Data::Object::Try->new(
          invocant => Example->new,
          arguments => [1,2,3]
        );
      
        $try->call(fun (@args) {
      
          return $try;
        });
      
        $try->finally(fun (@args) {
      
          $try->{'$finally'} = [@args];
        });

 maybe

      maybe() : Object

    The maybe method registers a default catch condition that returns
    falsy, i.e. an empty string, if an exception is encountered.

    maybe example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          die $try;
        });
      
        $try->maybe;

 no_catch

      no_catch() : Object

    The no_catch method removes any configured catch conditions and returns
    the object.

    no_catch example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          die $try;
        });
      
        $try->catch('Data::Object::Try', fun (@args) {
      
          return [@args];
        });
      
        $try->no_catch;

 no_default

      no_default() : Object

    The no_default method removes any configured default condition and
    returns the object.

    no_default example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          die $try;
        });
      
        $try->default(fun (@args) {
      
          return [@args];
        });
      
        $try->no_default;

 no_finally

      no_finally() : Object

    The no_finally method removes any configured finally condition and
    returns the object.

    no_finally example #1

        my $try = Data::Object::Try->new(
          invocant => Example->new,
          arguments => [1,2,3]
        );
      
        $try->call(fun (@args) {
      
          return $try;
        });
      
        $try->finally(fun (@args) {
      
          $try->{'$finally'} = [@args];
        });
      
        $try->no_finally;

 no_try

      no_try() : Object

    The no_try method removes any configured try operation and returns the
    object.

    no_try example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          return [@args];
        });
      
        $try->no_try;

 result

      result(Any @args) : Any

    The result method executes the try/catch/default/finally logic and
    returns either 1) the return value from the successfully tried
    operation 2) the return value from the successfully matched catch
    condition if an exception was thrown 3) the return value from the
    default catch condition if an exception was thrown and no catch
    condition matched. When invoked, the try and finally callbacks will
    received an invocant if one was provided to the constructor, the
    default arguments if any were provided to the constructor, and whatever
    arguments were passed directly to this method.

    result example #1

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          return [@args];
        });
      
        $try->result;

    result example #2

        my $try = Data::Object::Try->new;
      
        $try->call(fun (@args) {
      
          return [@args];
        });
      
        $try->result(1..5);

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/data-object-try/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/data-object-try/wiki>

    Project <https://github.com/iamalnewkirk/data-object-try>

    Initiatives <https://github.com/iamalnewkirk/data-object-try/projects>

    Milestones <https://github.com/iamalnewkirk/data-object-try/milestones>

    Contributing
    <https://github.com/iamalnewkirk/data-object-try/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/data-object-try/issues>



CHANGES
Changelog for Data-Object-Try

2.01 2020-02-29T04:40:08
 - Added release updates
 - Updated documentation
 - Fix documentation and non-testing prereqs

2.00 2020-02-29T04:21:02
 - Built release version 2.00
 - Added release updates
 - Initial commit


lib/Data/Object/Try.pm
package Data::Object::Try;

use 5.014;

use strict;
use warnings;
use routines;

use Moo;
use Try::Tiny ();

# VERSION

# ATTRIBUTES

has 'invocant' => (
  is => 'ro'
);

has 'arguments' => (
  is => 'ro'
);

has 'on_try' => (
  is => 'rw'
);

has 'on_catch' => (
  is => 'rw'
);

has 'on_default' => (
  is => 'rw'
);

has 'on_finally' => (
  is => 'rw'
);

# BUILD

method BUILD($args) {
  $self->{'on_catch'} = [] if !$args->{'on_catch'};

  return $args;
}

# METHODS

method call($callback) {
  $self->on_try($self->callback($callback));

  return $self;
}

method callback($callback) {
  require Carp;

  unless (UNIVERSAL::isa($callback, 'CODE')) {
    my $method = $self->invocant
      ? $self->invocant->can($callback) : $self->can($callback);
      Carp::confess(sprintf(
        qq(Can't locate object method "%s" on package "%s"),
        ($callback, ref $self)
      )) if !$method;
    $callback = sub { goto $method };
  }

  return $callback;
}

method catch($class, $callback) {
  push @{$self->on_catch}, [$class, $self->callback($callback)];

  return $self;
}

method default($callback) {
  $self->on_default($self->callback($callback));

  return $self;
}

method execute($callback, @args) {
  unshift @args, @{$self->arguments}
    if $self->arguments && @{$self->arguments};
  unshift @args, $self->invocant
    if $self->invocant;

  return $callback->(@args);
}

method finally($callback) {
  $self->on_finally($self->callback($callback));

  return $self;
}

method maybe() {
  $self->on_default(sub{''});

  return $self;
}

method no_catch() {
  $self->on_catch([]);

  return $self;
}

method no_default() {
  $self->on_default(undef);

  return $self;
}

method no_finally() {
  $self->on_finally(undef);

  return $self;
}

method no_try() {
  $self->on_try(undef);

  return $self;
}

method result(@args) {
  require Carp;

  my $returned;

  Try::Tiny::try(sub {
    my $tryer = $self->on_try;

    $returned = $self->execute($tryer, @args);
  }, Try::Tiny::catch(sub {
    my $caught = $_;
    my $catchers = $self->on_catch;
    my $default = $self->on_default;

    for my $catcher (@$catchers) {
      if (UNIVERSAL::isa($caught, $catcher->[0])) {
        $returned = $catcher->[1]->($caught);
        last;
      }
    }

    if(!$returned) {
      $returned = $default->($caught) if $default;
      Carp::confess($caught) if not defined $returned;
    }
  }, Try::Tiny::finally(sub {
    my $finally = $self->on_finally;

    $self->execute($finally, @args) if $finally;
  })));

  return $returned;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Try Class for Perl 5",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Data-Object-Try",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Data::Object::Try",
  "PREREQ_PM" => {
    "Moo" => 0,
    "Test::Auto" => "0.05",
    "Try::Tiny" => "0.30",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Moo" => 0,
    "Test::Auto" => "0.05",
    "Try::Tiny" => "0.30",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "2.00",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Moo" => 0,
  "Test::Auto" => "0.05",
  "Try::Tiny" => "0.30",
  "routines" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
